sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala

```diff
@@ -586,9 +584,139 @@ object ContainsExpressionBuilder extends StringBinaryPredicateExpressionBuilderB
 }
 
 case class Contains(left: Expression, right: Expression) extends StringPredicate {
-  override def compare(l: UTF8String, r: UTF8String): Boolean = l.contains(r)
+  override def inputTypes: Seq[DataType] = {
+    // TODO: properly handle String Types with differing collations
+    val leftIsString = left.dataType.getClass.getName.startsWith(
+      "org.apache.spark.sql.types.StringType")
+    val rightIsString = right.dataType.getClass.getName.startsWith(
+      "org.apache.spark.sql.types.StringType")
+    if (leftIsString && rightIsString) {
+      Seq(left.dataType, right.dataType)
+    } else {
+      Seq(StringType, StringType)
+    }
+  }
+
+  override def compare(l: UTF8String, r: UTF8String): Boolean = {
+    val lCollationID = left.dataType.asInstanceOf[StringType].collationId
+    val rCollationID = right.dataType.asInstanceOf[StringType].collationId
+    // TODO: ensure that lCollationID and rCollationID are equal (compatible)
+    val collationID = if (lCollationID > rCollationID) lCollationID else rCollationID
+    val collatorInfo = CollatorFactory.getInfoForId(collationID)
+    val collator = collatorInfo.collator
+    collator match {
+      // UTF8String collation
+      case null =>
+        collationID match {
+          // UCS_BASIC
+          case 0 =>
+            l.contains(r)
```

Comment:
Note: this is the original implementation before adding collation support (works only for UTF8String as is).

Reply:
Can we build a common helper in `CollationFactory` that will solve this? I was thinking about something like `getNormalized` which will return `collationKey`/`toLower` representations.

Also, let's not use 0, 1... but instead let's give some meaningful names. E.g. (pseudo code)
```scala
if collationId == DEFAULT_COLLATION_ID {
  l.contains(r) 
} else { 
  val coll = CollationFactory.fetchCollation(collationId)
  coll.normalize(left).contains(coll.normalize(right)
}

Reply:
I think this looks good! Maybe even just:
`if collationId == DEFAULT_COLLATION_ID {
  l.contains(r) 
} else { 
  CollationFactory.fetchCollation(collationId).contains(left, right)
}`

and then let `coll` take care of the rest

Reply:
Update: I think it looks better in the new [PR](https://github.com/apache/spark/pull/45216/files)